File: data-structures/lists/python/nested_lists_get_craft.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0


# Index values of each astronaut
CRAFT_INDEX = 0
NAME_INDEX = 1
DAYS_INDEX = 2


def get_astronauts_by_spacecraft(all_astronauts, given_spacecraft):
    """Return a list of astronauts that are on a given spacecraft"""

    # Declare a list to store the found astronauts in
    found_list = []

    # Repeat for all astronauts in the outer list using index i
    for i in range(len(all_astronauts)):
        # Check if the astronaut is in the spacecraft
        if given_spacecraft == all_astronauts[i][CRAFT_INDEX]:
            found_list.append(all_astronauts[i])

    return found_list
    

def display_all_astronauts(all_astronauts):
    """Display the values of each astronaut in the nested list"""

    # Repeat for each astronaut in the outer list
    for astronaut in all_astronauts:
        # Repeat for each value in the inner list
        for value in astronaut:
            print(f"{value}  ", end = "") # Do not print new line
        print() # Print a new line after each astronaut


def main():
    # Declare and initialise the nested list of astronauts
    all_astronauts = [
        ["ISS", "Sergey Prokopyev", "176"],
        ["ISS", "Dmitry Petelin", "176"],
        ["ISS", "Frank Rubio", "176"],
        ["Tiangong", "Fei Junlong", "112"],
        ["Tiangong", "Deng Qingming", "107"],
        ["Tiangong", "Zhang Lu", "107"],
        ["ISS", "Stephen Bowen", "54"],
        ["ISS", "Warren Hoburg", "14"],
        ["ISS", "Sultan Al Neyadi", "14"],
        ["ISS", "Andrey Fedyaev", "14"]
    ]
   
    # Find all the astronauts on a spacecraft
    print("Get a list of all astronauts that are on a specific spacecraft")
    spacecraft = input("Enter the spacecraft name: ")
    found_astronauts = get_astronauts_by_spacecraft(all_astronauts, spacecraft)
    
    # Output the search results
    if len(found_astronauts) == 0:
        print(f"\nNo astronauts found in the spacecraft {spacecraft}")
    else:
        print("\nAstronauts found:")
        display_all_astronauts(found_astronauts)


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()

File: data-structures/lists/c-sharp/nested_lists_get_craft.cs
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science 
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;
using System.Collections.Generic;

namespace AdaCodeSamples {
    class NestedLists {
        // Index values of each astronaut
        public const int CraftIndex = 0;
        public const int NameIndex = 1;
        public const int DaysIndex = 2;

        // The Main method is the entry point for all C# programs
        public static void Main() {
            // Declare and initialise the nested list of astronauts
            List<List<string>> allAstronauts = new List<List<string>>() {
                new List<string>() { "ISS", "Sergey Prokopyev", "176" },
                new List<string>() { "ISS", "Dmitry Petelin", "176" },
                new List<string>() { "ISS", "Frank Rubio", "176" },
                new List<string>() { "Tiangong", "Fei Junlong", "112" },
                new List<string>() { "Tiangong", "Deng Qingming", "107" },
                new List<string>() { "Tiangong", "Zhang Lu", "107" },
                new List<string>() { "ISS", "Stephen Bowen", "54" },
                new List<string>() { "ISS", "Warren Hoburg", "14" },
                new List<string>() { "ISS", "Sultan Al Neyadi", "14" },
                new List<string>() { "ISS", "Andrey Fedyaev", "14" }
            };

            // Find all the astronauts on a spacecraft
            Console.WriteLine("Get a list of all astronauts that are on a specific spacecraft");
            Console.WriteLine("Enter the spacecraft name:");
            string spacecraft = Console.ReadLine();
            List<List<string>> foundAstronauts = GetAstronautsBySpacecraft(allAstronauts, spacecraft);

            // Output the search results
            if (foundAstronauts.Count == 0) {
               Console.WriteLine($"\nNo astronauts found in the spacecraft {spacecraft}");
            }
            else {
               Console.WriteLine("\nAstronauts found:");
               DisplayAllAstronauts(foundAstronauts);
            }
        }
        

        // Return a list of astronauts that are on a given spacecraft
        public static List<List<string>> GetAstronautsBySpacecraft(
            List<List<string>> allAstronauts, string givenSpacecraft) {

            // Declare a list to store the found astronauts in
            List<List<string>> foundList = new List<List<string>>();

            // Repeat for all astronauts in the outer list using the index i
            for (int i = 0; i < allAstronauts.Count; i++) {
                // Check if the astronaut is in the spacecraft
                if (givenSpacecraft.Equals(allAstronauts[i][CraftIndex])) {
                    foundList.Add(allAstronauts[i]);
                }
            }

            return foundList;
        }


        // Display the values of each astronaut in the given list
        public static void DisplayAllAstronauts(List<List<string>> allAstronauts) {

            // Repeat for each astronaut in the outer list
            foreach (List<string> astronaut in allAstronauts) {
                // Repeat for each value in the inner list
                foreach (string value in astronaut) {
                    Console.Write($"{value}  "); // Do not print new line
                }
                Console.WriteLine(); // Print a new line after each astronaut
            }
        }
    }
}